torchx/schedulers/ray/ray_driver.py

```diff
@@ -147,7 +173,23 @@ def main() -> None:  # pragma: no cover
         # If a failure occurs the ObjectRef will be marked as completed.
         # Calling ray.get will expose the failure as a RayActorError.
         for object_ref in completed_workers:
-            ray.get(object_ref)
+            result = ray.get(object_ref)
+            if isinstance(result, PlacementGroup):
```

Comment:
can you explain how this works? not super familiar with the API but seems like we shouldn't be creating workers after some replicas have already finished?

Reply:
1. Once you call a remote function like `obj = fn.remote()`, the return value `obj` is just an object reference, ithe function `fn` may haven't been executed yet. You get the value of this reference only if you call `ray.get()`, and block until the function has been completed.
2. `ray.util.placement_group(bundle)`, bundle is a list of dict include the resources needed for this placement group. And this call is non-blocking, you get the reference of the placement group directly without waiting. The placement group can be created or still pending, but they all have a reference. `obj = pg.ready()` is actually a *remote call* to test if the placement group has been created, and it's also an object reference, you have to use `ray.get(obj)` to get the return value(`Optional[PlacementGroup]`, where `PlacemenGroup` is the reference to the placement group just mentioned) of `pg.ready()`.
3. `actor.exec_module.remote()` is a remote call that let the command actor run the script the user provided. It doesn't return any value, but you still have an object reference for `None`. (Every `ray.remote` call) returns an object reference, you only get the value once you use `ray.get`)
4. Since both `actor.exec_module.remote()` and `pg.ready()`  return object references, we can use `ray.wait()` to wait on both of them. `ray.wait` is blocking and partial waiting, as long as one of the function has completed, it returns the completed task and the tasks are still waiting.
5. The logic in the loop is we think both placement group creation and command actor execution as a task, based on the type of its return value, we know if it's an actor completed or a new placement group is created. If it's a placement group created, we can put a command actor on it.

Reply:
Thanks! I missed the pg.ready() bit, sounds reasonable. The docstring doesn't need to include quite this much detail just the high level

Reply:
I am also still a little reluctant on the fact that both the pg tasks and actor tasks are going into the same list and we are using instance type to separate out what logic we should be doing.

`completed_tasks, active_tasks = ray.wait(active_tasks)` now means that both pg and the actor will wait for the same period of time. 

there also doesn't seem be any exit out if the pg creation never happens, the loop will keep infinitely running even if the training has been completed. 

I like the idea of creating pg via tasks, but I think trying to accomplish everything in the same loop os introducing too many complications and we be better off using asyncio and treat these as 2 separate loops running in parallel

Reply:
@atinsood Actually, those tasks that from the command actors will never be finished until the entire training process is finished. The command actors' tasks will always return a None, while the placement groups will always return a reference to the placement group.
```
            result = ray.get(object_ref)  # pyre-ignore
            if isinstance(result, PlacementGroup):
                new_actor: CommandActor = create_command_actors(
                    [
                        actors[
                            pg_ids[result.id]
                        ],  # find the actor of a placement group based on pg_id
                    ],  # must be a list
                    result,
                )[0]
                active_tasks.append(
                    new_actor.exec_module.remote()  # pyre-ignore
                )  # add a command actor execution task
            else:
                # if the result is None, it's returned by a completed worker, and the job is completed
                return
```
There is an `else` branch, which means if the result is None(non-placement group). it is returned by a command actor, and it means the entire training process is finished, then the return will end this process.

ray.wait on the command actors doesn't do anything but wait the actors to finish their jobs. Just like waiting for the child process to be completed.